jina/peapods/runtimes/grpc/__init__.py

```diff
@@ -0,0 +1,266 @@
+from typing import Optional, Union
+
+from grpc import RpcError
+
+from jina.enums import OnErrorStrategy
+from jina.excepts import NoExplicitMessage, ChainedPodException, RuntimeTerminated
+from jina.helper import get_or_reuse_loop, random_identity
+from jina.peapods.grpc import Grpclet
+from jina.peapods.runtimes.base import BaseRuntime
+from jina.peapods.runtimes.request_handlers.data_request_handler import (
+    DataRequestHandler,
+)
+from jina.proto import jina_pb2
+
+from jina.types.routing.table import RoutingTable
+
+
+class GRPCDataRuntime(BaseRuntime, ABC):
+    """Runtime procedure leveraging :class:`Grpclet` for sending DataRequests"""
+
+    def __init__(self, args: 'argparse.Namespace', **kwargs):
+        """Initialize grpc and data request handling.
+        :param args: args from CLI
+        :param kwargs: extra keyword arguments
+        """
+        super().__init__(args, **kwargs)
+        self._id = random_identity()
+        self._loop = get_or_reuse_loop()
+        self._last_active_time = time.perf_counter()
+
+        self._pending_msgs = defaultdict(list)  # type: Dict[str, List['Message']]
+        self._partial_requests = None
+
+        self._data_request_handler = DataRequestHandler(args, self.logger)
+        self._grpclet = Grpclet(
+            async_loop=self._loop,
+            args=self.args,
+            message_callback=self._callback,
+            logger=self.logger,
+        )
+
+    def run_forever(self):
+        """Start the `Grpclet`."""
+        self.logger.debug('Run GRPCDataRuntime')
+        self._grpclet_task = self._loop.create_task(self._grpclet.start())
+        try:
+            self._loop.run_until_complete(self._grpclet_task)
+        except CancelledError:
+            self.logger.debug('GRPCDataRuntime cancelled')
+            raise
+
+    def teardown(self):
+        """Close the `Grpclet` and `DataRequestHandler`."""
+        self.logger.debug('Teardown GRPCDataRuntime')
+
+        self._data_request_handler.close()
+        self._loop.stop()
+        self._loop.close()
+
+        super().teardown()
+
+    def _close_grpclet(self):
+        self._loop.create_task(self._grpclet.close())
+        self._grpclet_task.cancel()
+
+    @staticmethod
+    def get_control_address(**kwargs):
+        """
+        Does return None, exists for keeping interface compatible with ZEDRuntime
+
+        :param kwargs: extra keyword arguments
+        :returns: None
+        """
+        return None
+
+    @staticmethod
+    def is_ready(ctrl_address: str, **kwargs) -> bool:
+        """
+        Check if status is ready.
+
+        :param ctrl_address: the address where the control message needs to be sent
+        :param kwargs: extra keyword arguments
+
+        :return: True if status is ready else False.
+        """
+
+        try:
+            response = Grpclet.send_ctrl_msg(ctrl_address, 'STATUS')
+        except RpcError:
+            return False
+
+        return True
```

Comment:
it should check the value of the response? Not sure about this

Reply:
With this mechanism there is no real response. All gRPC calls return some empty protobuf struct, which is always the same. So the only thing 'STATUS' actually test here is that we can do the gRPC call. So if some RpcError occurs (no connection etc) we know that status is not ready. Otherwise we assume its okay.

Reply:
But for `status` we may want to provide different results right? Maybe not now, but it can be important in the future?